### PR TITLE
ISPN-9416 Support duplicates in Multimap

### PR DESCRIPTION
**Pull request by miador**
https://issues.redhat.com/browse/ISPN-9416

### CODE REVIEW

File: multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCache.java
Comment (karesti) [inline multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCache.java:146]: this has not been changed

File: multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCache.java
Comment (karesti) [inline multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCache.java:165]: no change required

File: multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCacheManager.java
Comment (karesti) [inline multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCacheManager.java:39]: pass the boolean

**Review by karesti — CHANGES_REQUESTED** [review]
small changes and we need to change the hotrod client too

**Review by wburns — CHANGES_REQUESTED** [review]
The `remove` method needs to also be updated. According to the documentation it states `In the case where duplicates are supported, all the key-value pairs will be removed.` So we need to make sure to abide by that.

This also brings up do we want to expose other methods such as `count(K, V)` or the like that returns how many values are mapped to the same key?

**Comment by miador** [comment]
For testing purposes, i've put 2 methods into `Codec31`. Once [ISPN-13977](https://github.com/infinispan/infinispan/pull/10196) is merged, i'll move them to `Codec40`

**Comment by wburns** [comment]
> For testing purposes, i've put 2 methods into `Codec31`. Once [ISPN-13977](https://github.com/infinispan/infinispan/pull/10196) is merged, i'll move them to `Codec40`

The PR is in main, if you could rebase and update 👍

**Comment by karesti** [comment]
@miador needs rebase :)

**Comment by miador** [comment]
> @miador needs rebase :)

Done 👍

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/RemoteMultimapCacheImpl.java
Comment (wburns) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/RemoteMultimapCacheImpl.java:66]: Lets change this so this method just invokes the other constructor. This way we don't need to duplicate the tracing or assignment of variables.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/RemoteMultimapCacheImpl.java
Comment (wburns) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/RemoteMultimapCacheImpl.java:49]: This can be final and moved to the other final variables.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/multimap/MultimapCacheManager.java
Comment (wburns) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/multimap/MultimapCacheManager.java:15]: We should document this method as returning a multimap that doesn't support duplicates.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/multimap/RemoteMultimapCacheAPITWithDuplicatesTest.java
Comment (wburns) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/multimap/RemoteMultimapCacheAPITWithDuplicatesTest.java:36]: You can just use the class name of the test for the cache name.

File: multimap/src/main/java/org/infinispan/multimap/api/embedded/MultimapCacheManager.java
Comment (wburns) [inline multimap/src/main/java/org/infinispan/multimap/api/embedded/MultimapCacheManager.java:28]: Same thing regarding documentation that the returned multi map cache does not support duplicates.

File: multimap/src/main/java/org/infinispan/multimap/impl/Bucket.java
Comment (wburns) [inline multimap/src/main/java/org/infinispan/multimap/impl/Bucket.java:84]: This code is not correct for a few different reasons. But this code can be simpler.

```java
         while (it.hasNext()){
            V v = it.next();
            if (Objects.deepEquals(v, value)) {
               removed = true;
            } else {
               newBucket.add(v);
            }
         }
```

And the `add remaining values` comment and code snippet can be put in the other else block as this code doesn't need it.

File: multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCache.java
Comment (wburns) [inline multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCache.java:49]:
```suggestion
 * MultimapCache can optionally supports duplicate values on keys. {@link
```

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/multimap/RemoteMultimapCacheAPITWithDuplicatesTest.java
Comment (miador) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/multimap/RemoteMultimapCacheAPITWithDuplicatesTest.java:36]: So it would be `RemoteMultimapCacheTest`?

**Review by karesti — APPROVED** [review]
LGTM!!! 
Great first contribution, congratulations!
Please, squash the commits in a single one containing the issue name and the description

File: multimap/src/test/java/org/infinispan/multimap/impl/EmbeddedMultimapCacheWithDuplicatesTest.java
Comment (wburns) [inline multimap/src/test/java/org/infinispan/multimap/impl/EmbeddedMultimapCacheWithDuplicatesTest.java:42]: I am personally not a fan of the multiple chained stages for these tests, as I think it is a lot easier to understand with individual awaits. The individual awaits also make the stack trace a lot easier to parse when a failure happens. But I would say it is okay to integrate as it.

**Review by wburns — CHANGES_REQUESTED** [review]
Also sorry I didn't catch this earlier, but all of these changes need to go into the client/hotrod module as well.

**Review by wburns — APPROVED** [review]
LGTM, waiting on CI before merging.

**Comment by miador** [comment]
Hmm, after i squashed commits, CI aborted the tests. Seems it had 
```bash
[2022-07-12T08:21:21.440Z] Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "timeout-thread--p34-t1"
[2022-07-12T08:21:21.440Z] Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "jgroups-16,rhos-infinispan-ci-1-33120"
```
I didn't change anything though.

**Comment by wburns** [comment]
> Hmm, after i squashed commits, CI aborted the tests. Seems it had
> 
> ```shell
> [2022-07-12T08:21:21.440Z] Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "timeout-thread--p34-t1"
> [2022-07-12T08:21:21.440Z] Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "jgroups-16,rhos-infinispan-ci-1-33120"
> ```
> 
> I didn't change anything though. Is it related to CI server?

I am looking at the heap dump, it seems unrelated to your changes. However, I started a new build in the meantime.

**Comment by miador** [comment]
> > Hmm, after i squashed commits, CI aborted the tests. Seems it had
> > ```shell
> > [2022-07-12T08:21:21.440Z] Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "timeout-thread--p34-t1"
> > [2022-07-12T08:21:21.440Z] Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "jgroups-16,rhos-infinispan-ci-1-33120"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I didn't change anything though. Is it related to CI server?
> 
> I am looking at the heap dump, it seems unrelated to your changes. However, I started a new build in the meantime.

Oh cool! It seems it built without problems :)

**Comment by wburns** [comment]
Integrated into main, thanks @miador !

**Pull request merged**
